### PR TITLE
Add frameworking for development board presets

### PR DESCRIPTION
**Pull request by zvecr**
## Description

Cherry picked whats ready to go from https://github.com/qmk/qmk_firmware/pull/16598.

This PR aims to...
* Provide context for later conversion logic to key against
* Reduce boilerplate
* Later enhance new-keyboard?

Name is still up for grabs as "development_board" still feels a little wordy. (Currently chosen as it seems to be common enough when browsing various documention/stores). Alternative names?
* "preset"
* "carrier_board"
* "dev_board"
* "controller"
* "module" - maybe too close to RIOT terminology

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by KarlK90** [comment]
Development board is ubiquitous but I would prefer `controller` in QMK, `preset` is to generic in my opinion.

**Review by KarlK90 — APPROVED** [review]

File: lib/python/qmk/info.py [hunk 1]
```diff
@@ -472,6 +473,18 @@ def _extract_config_h(info_data):
...
 
+def _process_defaults(info_data):
+    """Proccess any additional defaults based on currently discovered infomation
+    """
+    defaults_map = json_load(Path('data/mappings/defaults.json'))
+    for default_type in defaults_map.keys():
+        thing_map = defaults_map[default_type]
+        if default_type in info_data:
+            for key, value in thing_map.get(info_data[default_type], {}).items():
+                info_data[key] = value
```
Comment (Erovia) [inline lib/python/qmk/info.py:484]:
```suggestion
                if key not in info_data:
                    info_data[key] = value
```

We probably don't want to overwrite the data from the `info.json` with the defaults.

**Comment by Erovia** [comment]
> Development board is ubiquitous but I would prefer `controller` in QMK, `preset` is to generic in my opinion.

I'm torn between `controller` and `development_board`. For some reason I prefer the latter, but wouldn't mind either, tbh.

File: lib/python/qmk/info.py [hunk 2]
```diff
@@ -473,6 +474,18 @@ def _extract_config_h(info_data, config_c):
     return info_data
 
 
+def _process_defaults(info_data):
+    """Proccess any additional defaults based on currently discovered infomation
```
Comment (tzarc) [inline lib/python/qmk/info.py:478]:
```suggestion
    """Process any additional defaults based on currently discovered information
```

**Review by tzarc — CHANGES_REQUESTED** [review]

**Review by tzarc — APPROVED** [review]

**Pull request merged**